Documentation for Kubernetes Provider (#2640)
CenturyLinkCloud/PublicKB: Cloud Application Manager/Deploying Anywhere/kubernetes.md

### Create a Template Box

Go to ***Boxes -> New -> Template*** and select *Kubernetes* as type. Fill the parameters as usual. Mind that if you define any *Requirements* you also need to list those in a Deployment Policy as *Claims* in order to be able to deploy the Template Box using that Policy.

A Kubernetes Template Boxes can have multiple template files, currently only in YAML format, but CAM also supports **Jinja syntax with YAML**. You can change the order of template files by dragging and moving with the mouse, or long-pressing and dragging on a touch-screen device. At least one template file is required to deploy a Template Box.

It is also valid to have multiple resources defined in a single template file, separated by a line of three dashes, like in the snippet below:

```Yaml
...
            memory: 100Mi
        ports:
        - containerPort: 80
---
apiVersion: v1
kind: Service
metadata:
  name: frontend
...
```

If you want to deploy the same Template Box into the same namespace multiple times, you need to **include CAM Service ID** in the name of deployed resources, like:

```Yaml
apiVersion: v1
kind: Service
metadata:

>   name: frontend{{ service}}

Reviewer: ```suggestion
  name: frontend-{{ service}}
```

...
```

> ***New Template*** dialog can either:

Reviewer: ```suggestion
The ***New Template*** dialog offers you the following options:
```

> * Create a blank template by specifying its filename (please follow **.yaml* convention)

Reviewer: ```suggestion
* Create a blank template by specifying its filename (please follow **.yaml** convention)
```

Author: Should render as *.yaml in italic, that's the reason for the "missing" asterisk.

* Upload a template file from disk
* Import template files from remote URL

If the URL is a GitHub repository, the files from that directory are automatically imported, as long as the filename indicates a valid template type. If a template file with the same name already exists, it won't get overwritten, but duplicate names will block the deployment, so you need to delete old version template file beforehand.

> 

Reviewer: Add a template box screenshoot somewhere in this section

### Deploying Kubernetes Template Boxes

> When **Deploying** a Template Box, you need to select the a suitable Deployment Policy of Kubernetes kind.

Reviewer: ```suggestion
When **Deploying** a Template Box, you need to select a suitable Kubernetes Deployment Policy with claims matching all your template box requirements.
```

On deployment, CAM is going to parse the templates, first by executing Jinja instructions and variable substitutions, then parsing the YAML documents as resources. For every resource defined in the templates, CAM adds some labels and annotate them with Service ID, then submitting them one-by-one to the Kubernetes cluster, to create new resources, or update existing ones. If a resource already exists, CAM generates a patch against the previous version of the template (stored as an annotation on the Kubernetes resource), or the existing specification read from the Kubernetes cluster. (In some cases, the later approach could fail: for ex. if the resource was created or modified outside of CAM.)

> 

Reviewer: Add screenshot

### Endpoints

CAM collects a list of Ingresses from deployed resources having *loadBalancer* node in their statuses, and lists them on the *Endpoints* tab.

> Endpoints are collected right after deployment, or when reconfigured the instance.

Reviewer: Is this accurate? What about the polling to gather these ones?

Reviewer: Add screenshot

### Output Variables

> There is no way to define custom output variables in Kubernetes templates, but CAM is creating output variables from Endpoints, like *name*_IP, *name*_Kind, *name*_Port and *name*_Protocol.

Reviewer: ```suggestion
There is no way to define custom output variables in Kubernetes templates, but Cloud Application Manager will automatically create output variables from the available Endpoints, like ***name*_IP**, ***name*_Kind**, ***name*_Port** and ***name*_Protocol**.
```

> You can see output variables of deployed Instances in the ***Lifecycle Editor***.

Reviewer: ```suggestion
You can see the output variables of any deployed Instance in the ***Lifecycle Editor***.
```

Reviewer: Add screenshot

### List of Deployed Resources

> The resources that have been created by the deployment are going to populate the *Resources* tab of the Instance, and will be deleted when the instance is terminated.

Reviewer: ```suggestion
The resources that have been created by the deployment are shown in the *Resources* tab of the Instance details page and will be deleted when the instance is terminated.
```
Add screenshot

> Resources are collected right after deployment, or when reconfigured the instance.

Reviewer: ```suggestion
Resources are collected right after deployment, or when reconfiguring the instance.
```
(is this accurate? what about the polling?)

![Details of a new Kubernetes Template Box](../../images/cloud-application-manager/kubernetes/new_template_box_dialog.png)

A Kubernetes Template Box can have multiple template files, currently only in YAML format, but Cloud Application Manager also supports **Jinja syntax with YAML**. You can change the order of template files by dragging and dropping with the mouse, or long-pressing and dragging on a touch-screen device. At least one template file is required to deploy a Kubernetes Template Box.

![A Kubernetes Template Box with multiple templates](../../images/cloud-application-manager/kubernetes/template_box.png)

It is also valid to have multiple resources defined in a single template file, separated by a line of three dashes, like in the snippet below:

```Yaml
...
            memory: 100Mi
        ports:
        - containerPort: 80
---
apiVersion: v1
kind: Service
metadata:
  name: frontend
...
```

If you want to deploy the same Template Box into the same namespace multiple times, you need to **include Cloud Application Manager Service ID** in the name of the deployed resources, to avoid name collisions, like:

```Yaml
apiVersion: v1
kind: Service
metadata:

>   name: frontend-{{ service}}

Reviewer 2: `{{` must be commented or the page will show an error `\{{`
